src/vstest.console/Internal/MSBuildLogger.cs

```diff
@@ -57,39 +59,76 @@ public void Initialize(TestLoggerEvents events, string testRunDirectory)
         Output ??= ConsoleOutput.Instance;
 
         // Register for the events.
-        // events.TestRunMessage += TestMessageHandler;
+        events.TestRunMessage += TestMessageHandler;
         events.TestResult += TestResultHandler;
         events.TestRunComplete += TestRunCompleteHandler;
-        // events.TestRunStart += TestRunStartHandler;
-
-        // Register for the discovery events.
-        // events.DiscoveryMessage += TestMessageHandler;
     }
 
     public void Initialize(TestLoggerEvents events, Dictionary<string, string?> parameters)
     {
         Initialize(events, string.Empty);
     }
 
+    private void TestMessageHandler(object? sender, TestRunMessageEventArgs e)
+    {
+        switch (e.Level)
+        {
+            case TestMessageLevel.Informational:
+                SendMessage($"output-info", e.Message);
+                break;
+            case TestMessageLevel.Warning:
+                SendMessage($"output-warning", e.Message);
+                break;
+            case TestMessageLevel.Error:
+                SendMessage($"output-error", e.Message);
+                break;
+        }
+    }
+
     private void TestRunCompleteHandler(object? sender, TestRunCompleteEventArgs e)
     {
         TPDebug.Assert(Output != null, "Initialize should have been called");
 
         if (e.IsCanceled)
         {
-            Output.Error(false, CommandLineResources.TestRunCanceled);
+            SendMessage("run-cancel", CommandLineResources.TestRunCanceled);
         }
         else if (e.IsAborted)
         {
             if (e.Error == null)
             {
-                Output.Error(false, CommandLineResources.TestRunAborted);
+                SendMessage("run-abort", CommandLineResources.TestRunAborted);
             }
             else
             {
-                Output.Error(false, CommandLineResources.TestRunAbortedWithError, e.Error);
+                SendMessage("run-abort", string.Format(CultureInfo.CurrentCulture, CommandLineResources.TestRunAbortedWithError, e.Error));
             }
         }
+        else
+        {
+            var total = e.TestRunStatistics?.ExecutedTests ?? 0;
+            var passed = e.TestRunStatistics?[TestOutcome.Passed] ?? 0;
+            var skipped = e.TestRunStatistics?[TestOutcome.Skipped] ?? 0;
+            var failed = e.TestRunStatistics?[TestOutcome.Failed] ?? 0;
+            var time = e.ElapsedTimeInRunningTests.TotalMilliseconds;
+
+            var summary = string.Format(CultureInfo.CurrentCulture, CommandLineResources.TestRunSummary,
+                        (failed > 0 ? CommandLineResources.FailedTestIndicator : CommandLineResources.PassedTestIndicator) + "!",
+                        failed,
+                        passed,
+                        skipped,
+                        total,
+                        $"[{GetFormattedDurationString(e.ElapsedTimeInRunningTests)}]"
```

Comment:
This is not handling the null case.

Reply:
Which null case do you mean?

Reply:
GetFormattedXXX can return null. On the other check, you explicitely do if/else and when null you remove the [nullable] part. In here you don't handle the null part so we would end up with [] or [null]

Reply:
👍fixed and merged.